stat/distmv/normal.go

```diff
@@ -332,18 +332,81 @@ func NormalRand(dst, mean []float64, chol *mat.Cholesky, src rand.Source) []floa
 	return dst
 }
 
+// EigenSym is an eigendecomposition of a symmetric matrix.
+type EigenSym interface {
+	mat.Symmetric
+	// RawValues returns all eigenvalues in ascending order. The returned slice
+	// must not be modified.
+	RawValues() []float64
+	// RawQ returns an orthogonal matrix whose columns contain the eigenvectors.
+	// The returned matrix must not be modified.
+	RawQ() mat.Matrix
+}
+
+// PositivePartEigenSym is an EigenSym that sets any negative eigenvalues from
+// the given eigendecomposition to zero but otherwise returns the values
+// unchanged.
+//
+// This is useful for filtering eigenvalues of positive semi-definite matrices
+// that are almost zero but negative due to rounding errors.
+type PositivePartEigenSym struct {
+	ed   *mat.EigenSym
+	vals []float64
+}
+
+var _ EigenSym = (*PositivePartEigenSym)(nil)
+var _ EigenSym = (*mat.EigenSym)(nil)
+
+// NewFilteredEigenSym returns a new FilteredEigenSym, wrapping the given
+// eigendecomposition.
+func NewPositivePartEigenSym(ed *mat.EigenSym) *PositivePartEigenSym {
+	n := ed.SymmetricDim()
+	vals := make([]float64, n)
+	for i, lamda := range ed.RawValues() {
+		if lamda >= 0 {
```

Comment:
```suggestion
		if lamda > 0 {
```
? since `vals[i]` is already zero.

Reply:
Done.